### DISCUSSION TITLE
(Taiko) Volume changes happening too early sometimes

### DISCUSSION THREAD

**Discussion by chayleaf** [ppy/osu#15761] [Q&A]
~~In stable, people often use inherited timing points to create gradual volume changes, even while the swell is active. In lazer, swells instead override the hit volume, so instead of a gradual volume shift you will hear, for example, quiet sound while hitting the swell, and loud sound as soon as you cleared it.~~

~~To make the behavior match stable, hitsound volume shouldn't be affected by swells, instead the timing point's volume should be used.~~

Not sure what the cause of the issue is, but after certain swells there's an abrupt volume change, I thought it's caused by the swell overriding the hitsounds but that's apparently not the case, it might actually be the note overriding the hitsounds before it's even supposed to be hit?...

**Comment by bdach** [comment]
just for the sanity of everyone involved, can you link a map where this is used to test?

while yes, this will have to be a consideration given samples are assigned per object in lazer (and so "global" sample points no longer exist), i believe this should "just work" for legacy maps, unless there is some detail i am missing.

**Reply by chayleaf** [reply]
now that i looked into it the issue appears to not be related to swells overriding timing points, as there's no timing points mid-swell. That said, here's an example map https://osu.ppy.sh/beatmapsets/464643#/994817 (happens after the third swell at the very start)

Might be caused by the game using the hitsound volume from the next note too early, or something

**Reply by bdach** [reply]
your descriptions are too vague to understand conclusively. to make sure we're talking about the same issue: here's a video with three repeats of the linked map. on the last two there is an audible "snap" sort of sound on the first swell. is that what you are trying to report?

https://user-images.githubusercontent.com/20418176/143075429-48c99e2b-e4a2-4437-a6b7-b7692d353ab9.mp4

**Reply by chayleaf** [reply]
i already said i'm talking about the third swell. I'm talking about the fact the notes hit after the third swell but before the first note are too loud (auto won't help here)

**Reply by bdach** [reply]
> I'm talking about the fact the notes hit after the third swell but before the first note are too loud (auto won't help here)

it would have been way more helpful to state this from the start, or better yet, attach a comparison video, rather than me having to extract information piecemeal. it's extremely tiring after a while.

**Comment by bdach** [comment]
in any case, this is happening because of the following logic, introduced when global sample control points were removed in favour of per-hit-object ones.

https://github.com/ppy/osu/blob/216c18b0bb83158ff1e6ab3670777655b4fd3a7f/osu.Game/Rulesets/UI/GameplaySampleTriggerSource.cs#L66-L92

this logic is used by both taiko and mania, and works in a way wherein if the player is between objects, the taiko drum or the mania keys will use the sample point of the next object, rather than the previous one as was expected here. i'm not sure which is the correct choice here, seems like it would be situational. maybe some interpolation/tweening between the previous and next objects is in order here.

@peppy thoughts?

**Reply by chayleaf** [reply]
so i guess that means you can't control the volume *during* the swell after all, i understand taiko isn't exactly a priority but it's not that uncommon in taiko mapping to control hitsounds' volume even during empty sections or swells because of how common ghosttapping is (or in case of swells it serves an "actual purpose" even if the player doesn't ghosttap)... don't even know what the best option here is, changing the entire hitsound system for a single mode would probably be weird?...

**Reply by bdach** [reply]
the fact that the top-level objects receive just one sample point is not necessarily set in stone. we may be looking to provide a way to, for instance, specify per-node samples for things like osu! sliders (https://github.com/ppy/osu/issues/15618#issuecomment-968288164), or in the longer term, have something like tweening between sample points (https://github.com/ppy/osu/discussions/13141#discussioncomment-791670).
